### PR TITLE
Reorganize Gene Sets and implement PhenotypeDotPlot

### PR DESCRIPTION
**Pull request by GWMcElfresh**
Hi everyone, 

@bbimber has been mentioning that we should recreate some plots like these, so here's a first stab at leveraging the gene sets that we're collecting into Phenotyping.R (which I've slightly re-organized because part of this function will need manual updating)

The function MakePhenotypingDotPlot collects and organizes genes based on their `.RegisterGeneSet()` name, and if a gene is shared between multiple sets then it is organized into its own joint set.  Then, based on which gene sets the gene is placed into, it gets assigned a cell type. 

The plot is then faceted, and the first row of each facet label is the cell type, followed by GeneSets/phenotypes. 

Example:
image

let me know what you think!

### CODE REVIEW

**Review by bbimber — CHANGES_REQUESTED** [review]

File: R/Phenotyping.R [hunk 1]
```diff
@@ -348,4 +351,78 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+}
+
+#' @title MakePhenotypingDotPlot
+#'
+#' @description A helper function that Takes an input gene and identifies any entries matching registered gene sets. This function is the complement to GetGeneSet.
+#' @param yField The grouping variable used to calculate the average expression of genes and the y axis of the DotPlot.
+#' @param colorField Whether to color dots by scaled expression or unscaled expression.
+MakePhenotypingDotPlot <- function(seuratObj,
+                                   yField = 'ClusterNames_0.2',
+                                   colorField = 'scaled'
```
[the remark this answers is not in the document]
Comment (GWMcElfresh) [reply R/Phenotyping.R:382]: agreed, I'll change this.

File: R/Phenotyping.R [hunk 2]
```diff
@@ -348,4 +351,78 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+){
+  #Establish which gene sets will be queried  
+  meta_gene_list <- c(RIRA::GetGeneSet('TandNK_Activation.1'), RIRA::GetGeneSet('MemoryAndNaive'), RIRA::GetGeneSet('Cytotoxicity'), RIRA::GetGeneSet('ExhaustionOrInhibitory'), RIRA::GetGeneSet('Myelocytes'), RIRA::GetGeneSet('Pro_Myelocytes'))
```
[the remark this answers is not in the document]
Comment (GWMcElfresh) [reply R/Phenotyping.R:385]: Yeah, I've been iterating on this. I think that passing a list of gene set names into the function that get parsed and constructed into a "meta_gene_list" makes more sense. I'll update in the next commit.

File: R/Phenotyping.R [hunk 3]
```diff
@@ -348,4 +351,78 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+  dotplot_df <- plt$data
+  
+  #Begin Phenotyping
+  dotplot_df$Phenotype <- 0 #pass 0 as an easy debugging test for un/underannotated gene sets
+  matching_genesets <- .FindGeneList(dotplot_df$features.plot)
+  dotplot_df<- merge(dotplot_df, matching_genesets, by.x = "features.plot", by.y ='gene')
+  
+  #Parse Phenotype Field to determine which cell type the phenotype targets
+  dotplot_df$CellType <- 0 #pass 0 as an easy debugging test for un/underannotated gene sets
+  dotplot_df[grepl("MemoryAndNaive|EffectorCytokines|Exhaustion", dotplot_df$geneset_union), "CellType"] <- "Cell Type: T Cells"
```
[the remark this answers is not in the document]
Comment (GWMcElfresh) [reply R/Phenotyping.R:398]: Fair point about exhaustion. If there is overlap in specific gene sets between cell types, they should get collected into a joint cell type. At least that's the intention.

File: R/Phenotyping.R [hunk 4]
```diff
@@ -348,4 +351,78 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+    colorField <- sym("avg.exp.scaled")
+    colors <- c("blue", "white", "red")
+    colorLabel = 'Scaled Average Expression'
+  } else if(colorField == "unscaled") {
+    colorField <- sym("avg.expr")
+    colors = c("navy", "gold", "orange", "red")
+    colorLabel = 'Average Expression'
+  }
+  
+  P1 <- ggplot(dotplot_df, aes(x = features.plot, y = id, size = pct.exp, color = !!colorField)) +
```
[the remark this answers is not in the document]
Comment (GWMcElfresh) [reply R/Phenotyping.R:417]: it's a workaround for passing a string (e.g. `"avg.exp.scaled"` as a variable into dplyr/ggplot, so we can set the colorField based on whether the expression should be scaled or not. There's additionally a typo I need to fix before this functions as intended (avg.expr should be avg.exp).

File: R/Phenotyping.R [hunk 5]
```diff
@@ -348,4 +351,78 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
 	}
 
 	return(ret)
+}
+
+#' @title .FindGeneList
```
[the remark this answers is not in the document]
Comment (GWMcElfresh) [reply R/Phenotyping.R:356]: I agree. Maybe some kind of internal logic like `GeneList` refers to a vector of gene sets? (for `ExpandGeneList()`) and `GeneSets` refer to vectors of genes with an identifying gene name?

File: R/Phenotyping.R [hunk 6]
```diff
@@ -348,4 +351,84 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+#' @title MakePhenotypingDotPlot
+#'
+#' @description Creates a DotPlot using custom gene sets and attempts to coarsely group gene sets by cell type. 
+#' @param yField The grouping variable used to calculate the average expression of genes and the y axis of the DotPlot.
+#' @param scaled A boolean defining whether to color dots by scaled expression or unscaled expression.
+#' @param gene_lists A vector of gene lists (defined by .RegisterGeneSet) to be queried and their genes be plotted.
+MakePhenotypingDotPlot <- function(seuratObj,
+                                   yField = 'ClusterNames_0.2',
+                                   scaled = T, 
+                                   gene_lists = c('Cytotoxicity', 'EffectorCytokines')
```
Comment (GWMcElfresh) [inline R/Phenotyping.R:384]: @bbimber, I added support for supplying a vector of GeneSets. Do you have any gene sets that you'd want this to include by default? 

The warning from GetGeneSet provides some safety on nonsense input gene names.

File: R/Phenotyping.R [hunk 7]
```diff
@@ -348,4 +351,84 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+    colorLabel = 'Scaled Average Expression'
+  } else if(!scaled) {
+    colorField <- sym("avg.exp")
+    colors = c("navy", "gold", "orange", "red")
+    colorLabel = 'Average Expression'
+  }
+
+  P1 <- ggplot(dotplot_df, aes(x = features.plot, y = id, size = pct.exp, color = !!colorField)) + 
+    geom_point() + 
+    facet_wrap(~CellType+geneset_union, scales = "free_x") +
```
Comment (GWMcElfresh) [inline R/Phenotyping.R:425]: Additionally, I've been waffling on how to properly facet these. The GeneSet names are pretty long, so `facet_grid(...,scales = "free_x")` end up truncating the gene set. The alternative is sometimes rather empty panels, but everything is clearly legible. 

e.g.:
Screen Shot 2022-07-25 at 11 20 04 AM
 
 
 I think supporting custom `gene_lists` and then just interactively changing the `gene_lists` argument is good enough, but I'm definitely open to other ideas.

File: R/Phenotyping.R [hunk 8]
```diff
@@ -374,55 +374,63 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
 MakePhenotypingDotPlot <- function(seuratObj,
                                    yField = 'ClusterNames_0.2',
-                                   colorField = 'scaled'
+                                   scaled = T, 
+                                   gene_lists = c('Cytotoxicity', 'EffectorCytokines')
 ){
-  #Establish which gene sets will be queried  
-  meta_gene_list <- c(RIRA::GetGeneSet('TandNK_Activation.1'), RIRA::GetGeneSet('MemoryAndNaive'), RIRA::GetGeneSet('Cytotoxicity'), RIRA::GetGeneSet('ExhaustionOrInhibitory'), RIRA::GetGeneSet('Myelocytes'), RIRA::GetGeneSet('Pro_Myelocytes'))
+  if (!is.logical(scaled)){
+    warning("Please ensure scaled is either TRUE (to use per-gene scaled expression) or FALSE (for raw gene expression)")
```
Comment (bbimber) [inline R/Phenotyping.R:387]: I think you can just do:

stop("please ensure....")

instead of a warning and an empty stop()

**Review by bbimber — CHANGES_REQUESTED** [review]

File: R/Phenotyping.R [hunk 9]
```diff
@@ -348,4 +351,84 @@ ExpandGeneList <- function(genes, verbose = TRUE) {
...
+  return(unique_gene_set_pairs)
+}
+
+#' @title MakePhenotypingDotPlot
+#'
+#' @description Creates a DotPlot using custom gene sets and attempts to coarsely group gene sets by cell type. 
+#' @param yField The grouping variable used to calculate the average expression of genes and the y axis of the DotPlot.
+#' @param scaled A boolean defining whether to color dots by scaled expression or unscaled expression.
+#' @param gene_lists A vector of gene lists (defined by .RegisterGeneSet) to be queried and their genes be plotted.
+MakePhenotypingDotPlot <- function(seuratObj,
```
Comment (bbimber) [inline R/Phenotyping.R:381]: You probably want @export to this, right? you'll need to re-run devtools::document after that.

Comment (GWMcElfresh) [reply R/Phenotyping.R:382]: yeah, thank you!

File: R/Phenotyping.R [hunk 10]
```diff
@@ -430,5 +432,13 @@ MakePhenotypingDotPlot <- function(seuratObj,
     xlab("Genes") +
     labs(color=colorLabel, size = "Percentage of Cells\nWith Gene Expression")
   
+  scaling <- if(scale.by != "size" | scale.by != "radius"){
```
Comment (bbimber) [inline R/Phenotyping.R:435]: being picky, but it's probably easier to structure like:

if (size == 'x') {

} else if (size == 'y') {

} else {
    stop('.....')
}

the point being that you test the allowable values, and then die if it's not one of them. Easier to maintain than constantly adding to:
```
if(scale.by != "size" | scale.by != "radius"){
```

**Review by bbimber — CHANGES_REQUESTED** [review]

File: R/Phenotyping.R (hunk 10 above)
Comment (GWMcElfresh) [reply R/Phenotyping.R]: Agreed and changed

**Comment by bbimber** [comment]
@GWMcElfresh If the tests pass I think this is fine to merge

**Comment by GWMcElfresh** [comment]
The test that's failing seems to be scGate related? I did just add an assay passthrough (mostly to support nimble) so we'll see after this.

**Pull request merged**